File: Question 0139 - Word Break/src/Solution.java
import java.util.List;

class Solution {
    public boolean wordBreak(String s, List<String> wordDict)
    {
        Trie t = new Trie();
        for (int i = 0; i < wordDict.size(); i++) {
            t.insert(wordDict.get(i));
        }

        int n =s.length();
        boolean[] present = new boolean[n+1];
        present[0] = true;


        for (int i = 0; i < n + 1; i++) {
            for (int j = 0; j < i; j++) {
                if(present[j] && t.search(s.substring(j,i)))
                {
                    present[i] = true;
                    break;
                }
            }
        }

        return present[n];
    }
}

File: Question 0139 - Word Break/src/Solution.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution
{
private:
    bool solve(string &s, vector<string> &wordDict, int idx, vector<int> &dp)
    {
        // base case
        if (idx == s.size())
            return true;

        if (dp[idx] != -1)
            return dp[idx];

        for (auto &w : wordDict)
        {
            int len = w.size();

            if (s.substr(idx, len) == w)
            {
                if (solve(s, wordDict, idx + len, dp))
                    return dp[idx] = true;
            }
        }

        return dp[idx] = false;
    }

public:
    bool wordBreak(string s, vector<string> &wordDict)
    {
        vector<int> dp(s.size(), -1);
        return solve(s, wordDict, 0, dp);
    }
};